### ISSUE TITLE
_id undefined is sub schema array since 3.8.21

### ISSUE THREAD

**Issue by lcalvy** [LearnBoost/mongoose#2568]
Hi @vkarpov15,

Since this commit 1bf1e3b16916de2fc5b5da7e0ccae1179636a4a5 and code modification below, we lost the field _id from sub schema array after save.

` -      this._walkUpdatePath(ret, '$set');`
`+      this._walkUpdatePath(ret.$set || {}, '$set');`

For example, we have a mongoose schema like that
```javascript
var chapterSchema = {
    name: String
};
var schemaPrototype = {
        chapters: [chapterSchema],
        infiniteLives: {type: Boolean, default: false}
    };
```

Before this commit, object looks like
```javascript
{ _id: 54a2d064eb80a4a7b9f9b277,
  infiniteLives: false,
  chapters: 
   [ 
     { name: 'discipline_2.module_coach.chapter_1.name',
       _id: 54a2d064eb80a4a7b9f9b27c
     },
     { name: 'discipline_2.module_coach.chapter_2.name',
       _id: 54a2d064eb80a4a7b9f9b27b
      }
   ]
}
```
After this commit, object is

```javascript
{ _id: 54a2d064eb80a4a7b9f9b277,
   infiniteLives: false,
  chapters: 
   [ { name: 'discipline_2.module_coach.chapter_1.name'
       },
     { name: 'discipline_2.module_coach.chapter_2.name'
     }
   ]
}
```

Any idea?

regards

**Comment by vkarpov15** [comment]
Not sure, I'll look into it. Thanks for pointing this out.

**Comment by vkarpov15** [comment]
Hi @lcalvy, I'm going to need some more help to reproduce this. The below script:

```javascript
var mongoose = require('mongoose');
var assert = require('assert');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost:27017/gh2568');

var chapterSchema = {
  name: String
};
var schemaPrototype = {
  chapters: [chapterSchema],
  infiniteLives: {type: Boolean, default: false}
};

var Book = mongoose.model('gh2568', schemaPrototype, 'gh2568');

var b = new Book({
  infiniteLives: false,
  chapters: [{ name: 'Ursus' }, { name: 'The Comprachicos' }]
});

b.save(function(err) {
  assert.ifError(err);
  Book.findOne({ _id: b._id }, function(err, doc) {
    assert.ifError(err);
    console.log(JSON.stringify(doc, null, '  '));
    doc.chapters.push({ name: 'Portland Bill' });
    doc.save(function(err) {
      assert.ifError(err);
      Book.findOne({ _id: b._id }, function(err, doc) {
        assert.ifError(err);
        console.log(JSON.stringify(doc, null, '  '));
        process.exit(0);
      });
    });
  });
});
```

Outputs what I'd expect it to:

```
{
  "_id": "54da67686674b0361ca127c8",
  "__v": 0,
  "infiniteLives": false,
  "chapters": [
    {
      "name": "Ursus",
      "_id": "54da67686674b0361ca127ca"
    },
    {
      "name": "The Comprachicos",
      "_id": "54da67686674b0361ca127c9"
    }
  ]
}
{
  "_id": "54da67686674b0361ca127c8",
  "__v": 1,
  "infiniteLives": false,
  "chapters": [
    {
      "name": "Ursus",
      "_id": "54da67686674b0361ca127ca"
    },
    {
      "name": "The Comprachicos",
      "_id": "54da67686674b0361ca127c9"
    },
    {
      "name": "Portland Bill",
      "_id": "54da67686674b0361ca127cb"
    }
  ]
}
```

Could you provide me a standalone example that reproduces what you're seeing and/or some more insight as to how you're creating and saving docs?

**Comment by vkarpov15** [comment]
Can't repro and no response for a couple months. If this is still an issue, feel free to re-open.

**Comment by lcalvy** [comment]
Hi, After further test I found a repro case.
Issue occur when first model creation is done by an upsert command. 

```javascript
return Book.updateAsync({ref: file.content.ref}, file.content, {upsert: true, overwrite: true});
```

Where file.content is a pur JSON object.

**Comment by lcalvy** [comment]
@vkarpov15 

here a repro case

```javascript
'use strict';

var mongoose = require('mongoose');
var assert = require('assert');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost:27017/gh2568');

var chapterSchema = {
  name: String
};
var schemaPrototype = {
  chapters: [chapterSchema],
  infiniteLives: {type: Boolean, default: false},
  ref: {type: Number, unique: true}
};

var Book = mongoose.model('gh2568', schemaPrototype, 'gh2568');

var uniqueId = Date.now();

var jsonObject = {
  ref: uniqueId,
  infiniteLives: false,
  chapters: [{name: 'Ursus'}, {name: 'The Comprachicos'}]
};

Book.update({ref: 0}, jsonObject, {upsert: true, overwrite: true},
  function(err, book) {
      assert.ifError(err);
      console.log('OK');
      Book.findOne({ref: uniqueId}, function(err, book) {
        assert.ifError(err);
        assert.equal(book.chapters[0].name, 'Ursus');
        // should have one _id
        assert.notEqual(book.chapters[0]._id, undefined);
        console.log(book);
        process.exit(0);
      });
  });
```

**Comment by vkarpov15** [comment]
Thanks for the repro. Just pushed a fix and it'll be in for 3.8.25
